pkg/cluster/updater.go

```diff
@@ -0,0 +1,372 @@
+	// foreach(node in nodes)
+	// -- updateAnnotation(node)
+	// -- waitForUpdatedCurrentAnnotation(node, timeout)
+	var lo api.ListOptions
+	nl, err := client.Core().Nodes().List(lo)
+	if err != nil {
+		return err
+	}
+	for _, n := range nl.Items {
+		if n.Annotations == nil {
+			n.Annotations = make(map[string]string)
+		}
+		currentConfig := make(map[string]string)
+		err = json.Unmarshal([]byte(n.Annotations[node.CurrentConfigAnnotation]), &currentConfig)
+		if err != nil {
+			return err
+		}
+		currentConfig[node.KubeletVersionKey] = image.Tag
+		b, err := json.Marshal(currentConfig)
+		if err != nil {
+			return err
+		}
+		n.Annotations[node.DesiredConfigAnnotation] = string(b)
+		_, err = client.Core().Nodes().Update(&n)
+		if err != nil {
+			return err
+		}
+		lo := api.ListOptions{
+			Watch:         true,
+			FieldSelector: fields.ParseSelectorOrDie(fmt.Sprintf("metadata.name=%s", n.Name)),
+		}
+		w, err := client.Core().Nodes().Watch(lo)
+		if err != nil {
+			return err
+		}
+		for e := range w.ResultChan() {
+			switch e.Type {
+			case watch.Modified:
+				nn, ok := e.Object.(*v1.Node)
+				if !ok {
+					glog.Infof("recieved unexpected type from Node watch. Expected *v1.Node, got: %T", e.Object.GetObjectKind())
+					continue
+				}
+				if nn.Annotations[node.DesiredConfigAnnotation] == nn.Annotations[node.CurrentConfigAnnotation] {
+					break
+				}
+				glog.Infof("Node %s update completed", nn.Name)
+				break
+			case watch.Error:
+				aerr, ok := e.Object.(*unversioned.Status)
+				if !ok {
+					glog.Infof("recieved unexpected error type from Node watch. Expected *api.Status, got: %T", e.Object.GetObjectKind())
+					continue
+				}
+				glog.Errorf("recieved error during watch on Node %s: %s", n.Name, aerr.Message)
+			default:
+				glog.Infof("recieved event %s while watching node update, ignoring", e.Type)
+			}
+		}
+	}
+	return nil
+}
+
+// ClusterRollbackFunc is a function that can be used to
+// roll back an update to a cluster.
+type ClusterRollbackFunc func(clientset.Interface, *ContainerImage, *v1.ConfigMap) error
+
+// Update will update every component in the list of known components for
+// the instance of the ClusterUpdater. If any component should fail to update,
+// this method will automatically rollback the changes.
+func (cu *ClusterUpdater) Update() error {
+	for _, c := range cu.Components {
+		cu.Events.Eventf(cu.SelfObj, api.EventTypeNormal, "Begin update of component: %s", c.Name)
+		glog.Infof("Begin update of component: %s", c.Name)
+		if err := c.UpdateFn(cu.Client, c.Name, cu.NewVersion); err != nil {
+			cu.Events.Eventf(cu.SelfObj, api.EventTypeWarning,
+				"Failed update of component: %s, rolling back to version: %s from %s",
+				c.Name, cu.OldVersion, cu.NewVersion)
+			glog.Errorf("Failed update of component: %s, rolling back to version: %s from %s due to: %v",
+				c.Name, cu.OldVersion, cu.NewVersion, err)
+			return cu.Rollback()
```

Comment:
I think this should be an operator choice to rollback - not a rollback on any error. There are going to be transient network errors that the updateFn could hit, and we don't want to immediately roll back a cluster.